homeassistant/components/melcloud/__init__.py

```diff
@@ -66,7 +66,15 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Establish connection with MELClooud."""
     conf = entry.data
-    mel_devices = await mel_devices_setup(hass, conf[CONF_TOKEN])
+    try:
+        mel_devices = await mel_devices_setup(hass, conf[CONF_TOKEN])
+        _LOGGER.debug("Token used: %s", conf[CONF_TOKEN])
+    except (ClientResponseError, ConfigEntryNotReady) as ex:
+        _LOGGER.error("Failed to connect to MELCloud, with exception: %s", ex)
+        if isinstance(ex, ClientResponseError) and ex.code == 401:
+            raise ConfigEntryAuthFailed(ex) from ex
+        raise ConfigEntryNotReady from ex
```

Comment:
```suggestion
    except (ClientResponseError) as ex:
        if isinstance(ex, ClientResponseError) and ex.code == 401:
            raise ConfigEntryAuthFailed from ex
        raise ConfigEntryNotReady from ex
        except (asyncio.TimeoutError, ClientConnectionError) as ex:
            raise ConfigEntryNotReady from ex
```
If you change line 168-187 to:
```python
async def mel_devices_setup(
    hass: HomeAssistant, token: str
) -> dict[str, list[MelCloudDevice]]:
    """Query connected devices from MELCloud."""
    session = async_get_clientsession(hass)
    async with asyncio.timeout(10):
        all_devices = await get_devices(
            token,
            session,
            conf_update_interval=timedelta(minutes=5),
            device_set_debounce=timedelta(seconds=1),
        )
    wrapped_devices: dict[str, list[MelCloudDevice]] = {}
    for device_type, devices in all_devices.items():
        wrapped_devices[device_type] = [MelCloudDevice(device) for device in devices]
    return wrapped_devices
```

Reply:
Sure - I wasn't hesitant to not optimize too much, whereas that could be considered out-of-scope. Improvements pushed. :)